### DISCUSSION TITLE
CASE and BEGIN tag for SQL Query

### DISCUSSION THREAD

**Discussion by yanleiberg** [facelessuser/BracketHighlighter#612] [Q&A] [answered]
Hello,

I am requesting a feature to highlight the CASE END and BEGIN END braceket clauses for SQL queries. I have attempted to enable this feature using the settings provided below, but it is not working for me. 
Could you please provide assistance in getting this feature to work or suggest what is the proper setting should be?

Thank you.

```
{
    "scope_brackets": [
        // Quotes
        {
            "name": "sql_case",
            "open": "(?i)case",
            "close": "(?i)end",
            "style": "string",
            "scopes": ["string", "meta.string"],
            "language_filter": "allowlist",
            "language_list": [
                "Plain text"
            ],
            "sub_bracket_search": "true",
            "enabled": true
        },
        {
            "name": "sql_case",
            "open": "(?i)begin",
            "close": "(?i)end",
            "style": "string",
            "scopes": ["string", "meta.string"],
            "language_filter": "allowlist",
            "language_list": [
                "Plain text"
            ],
            "sub_bracket_search": "true",
            "enabled": true
        },
    ]
}
```

**Comment by facelessuser** [comment]
You would need to provide:

- the syntax language you are using (please link 3rd party syntax plugins)
- a syntax example to test the behavior as I do not know all languages

If you can provide the above, I'm happy to try out your attempt and advise accordingly.

**Comment by yanleiberg** [comment]
The syntax link:  
https://github.com/sublimehq/Packages/blob/master/SQL/SQL.sublime-syntax

An example for BEGIN ... END statement:  

```
BEGIN
    SELECT
        product_id,
        product_name
    FROM
        production.products
    WHERE
        list_price > 100000;

    IF @@ROWCOUNT = 0
        PRINT 'No product with price greater than 100000 found';
END
```

An example for CASE ... END statement:  

```
SELECT
    CustomerName
   ,City
   ,Country
FROM Customers
ORDER BY (CASE
	WHEN City IS NULL THEN 
	        CASE
			WHEN Country IS NULL THEN CustomerName
			ELSE Country
		END
	ELSE City
END);
```

The common characters that may appear before or after "BEGIN", "CASE" AND "END" are space, tab and linebreak.

The other characters that may appear before "BEGIN" are semicolon and left parenthesis.
The other characters that may appear after the "END" of "BEGIN END" are semicolon and right parenthesis.

The other character that may appear before "CASE" is left parenthesis.
The other character that may appear after the "END" of "CASE END" is right parenthesis.

The keywords like BEGIN, CASE, ELSE are case insensitive which means "begin", "BeGin" or "bEGIn" are also legitimate.

**Comment by facelessuser** [comment, marked as answer]
So, there are a number of issues wrong with what you tried:

1. You say you are using `SQL/SQL.sublime-syntax`, yet you are targeting `Plain text`. You will not match in SQL if you are targeting a syntax you are not using.

2. You are targeting `"scopes": ["string", "meta.string"],`, but those do not even match the scopes being used for this keyword terms. You need to target `keyword`.

3. You should be placing this in the `user_*` settings to add the new rules not replace the existing rules. `user_scope_brackets` and `user_brackets` is what you should be using.

4. You are targeting the `scope_brackets` rules which is used to target brackets that are hard to distinguish the start from the end. Think quotes whose start and end are identical. This is not needed with unique start and end brackets such as `CASE` and `END`, etc.

5. Both `CASE` and `BEGIN` end with `END`. To keep the wrong `END` from matching `CASE` or `BEGIN`, that context should be included in the rule. Better put, you need one rule for both `CASE` and `BEGIN`.

```js
    "user_brackets": [
        {
            "name": "sql_conditional",
            "open": "(?i)\\b(case|begin)\\b",
            "close": "(?i)(end)",
            "style": "string",
            "scopes": ["keyword"],
            "language_filter": "allowlist",
            "language_list": [
                "SQL"
            ],
            "enabled": true
        }
    ]
```

Now, keep in mind that there may be other keywords that end in `END`, so you might still have some bad matches until you finish resolving all of that with a proper rule. I don't know all the ends and outs of SQL, so you'll have to investigate more.

**Reply by yanleiberg** [reply]
This works! 
Thank you so much for your help! 
Your setting was exactly what I was looking for and has made a huge difference in my work. 


1 - I didn't see `SQL` as an syntax option in the Wrap Settings, so I thought I shouldn't use `SQL`.

2 - I just copied another setting and changed some fields. 

3 - Yes, it was placing in the user_* setting however the `user-scopt_brackets` option was not within my knowledge. 

4 - Same to anwser 2, copy paste issue + lack of the setting knowledge.  

5 - This is certainly more elegant! 

For the `END` keyword, it is not possible to have an `END` keyword in SQL without having a matching `BEGIN` or `CASE` keyword. 
The `END` keyword is used to mark the end of a `BEGIN...END` block or a `CASE...END` block, so it must always have a matching `BEGIN` or `CASE` keyword.

Thank you again for everything!

**Comment by yanleiberg** [comment]
One more question, I can only see this discusstion in the Discusstions tab, is this due to my privilege or this is indeed the only discussion?  
If it is related to my privilege, does that mean that other people are not able to see/find this solution? 

Thank you.

**Reply by facelessuser** [reply]
I moved your issue to Discussions, that's why it is now here. As this was a question, and not a feature request or a bug, it made more sense to move it Discussions as a Q & A.

**Reply by yanleiberg** [reply]
I totally understand for moving this to Discussions. 
My point is, is it normal that I see no other discussions at all exept this one?

**Reply by facelessuser** [reply]
I just enabled discussions for this repo. I apparently had not enabled it before now

**Reply by yanleiberg** [reply]
Thank you, that awnsers my question. 

Will this SQL setting be included directly in future version of BracketHighlighter, instead of adding it manually to user setting？

**Reply by facelessuser** [reply]
Since you are familiar with SQL, if you validate that is good to go and issue a pull request, we can consider it in the next release.

**Reply by yanleiberg** [reply]
I'll do more tests and come back to you, thank you.
